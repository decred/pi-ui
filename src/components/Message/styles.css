.message {
  padding: 20px 40px;
  display: flex;
  align-items: center;
}

.info {
  composes: message;
  background-color: var(--info-message-background);
  color: var(--info-message-color);
}

.icon-info {
  margin-right: 10px;
}

.icon-info > path:first-of-type {
  fill: var(--info-message-color);
}

.icon-info > path:last-of-type {
  fill: var(--info-message-background);
}

.error {
  composes: message;
  background-color: var(--error-message-background);
  color: var(--error-message-color);
}

.icon-error {
  margin-right: 10px;
}

.icon-error > path:first-of-type {
  fill: var(--error-message-color);
}

.icon-error > path:last-of-type {
  fill: var(--error-message-background);
}

.warning {
  composes: message;
  background-color: var(--warning-message-background);
  color: var(--warning-message-color);
}

.icon-warning {
  margin-right: 10px;
}

.icon-warning > path:first-of-type {
  fill: var(--warning-message-color);
}

.icon-warning > path:last-of-type {
  fill: var(--warning-message-background);
}

.success {
  composes: message;
  background-color: var(--success-message-background);
  color: var(--success-message-color);
}

.icon-success {
  margin-right: 10px;
}

.icon-success > path:first-of-type,
.icon-success > circle:first-of-type {
  fill: var(--success-message-color);
}

.icon-success > path:last-of-type {
  fill: var(--success-message-background);
}
