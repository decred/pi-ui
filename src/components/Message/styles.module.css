.message {
  padding: 20px 40px;
  display: flex;
  align-items: center;
  color: var(--text-color) !important;
}

.withTitle {
  align-items: flex-start;
}

.iconContainer {
  height: 100%;
  display: flex;
  align-items: center;
}

.withTitle .iconContainer {
  padding-top: 6px;
}

.iconForTitle {
  align-items: flex-start;
}

.content {
  display: flex;
  flex-direction: column;
}

.info {
  composes: message;
  background-color: var(--info-message-background);
}

.icon-info {
  margin-right: 10px;
}

.error {
  composes: message;
  background-color: var(--error-message-background);
}

.iconError {
  margin-right: 10px;
}

.iconError > path:first-of-type {
  fill: var(--error-message-color);
  width: 60px;
}

.iconError > path:last-of-type {
  fill: var(--error-message-background);
}

.warning {
  composes: message;
  background-color: var(--warning-message-background);
}

.iconWarning {
  margin-right: 10px;
}

.iconWarning > path:first-of-type {
  fill: var(--warning-message-color);
}

.iconWarning > path:last-of-type {
  fill: var(--warning-message-background);
}

.success {
  composes: message;
  background-color: var(--success-message-background);
}

.blocked {
  composes: message;
  background-color: var(--blocked-message-background);
}

.iconBlocked {
  margin-right: 10px;
}

.iconBlocked line,
.iconBlocked circle {
  fill: none;
  stroke: var(--blocked-message-color);
}

.iconSuccess {
  margin-right: 10px;
}

.iconSuccess > path:first-of-type,
.iconSuccess > circle:first-of-type {
  fill: var(--success-message-color);
}

.iconSuccess > path:last-of-type {
  fill: var(--success-message-background);
}
