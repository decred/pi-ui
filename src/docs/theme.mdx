---
name: Theming
route: /theming
---

# Theming

## Create a custom theme

The recommended way to create a custom theme is to create a file named `theme.js`, you can add your custom values and override existing ones.

```js
// theme.js

export default {
  "color-primary": "red",
  "btn-color": "var(--color-primary)"
};
```

## Use your theme in your application

Pi-ui exposes a `ThemeProvider` that accepts a `themes` object, `defaultThemeName` as string and `fonts` array.

```js
import React from "react";
import { render } from "react-dom";
import { ThemeProvider, defaultLightTheme } from "pi-ui";

import customTheme from "./theme";
import SourceSansProLight from "../assets/fonts/source-sans-pro/SourceSansPro-Light.ttf";
import SourceSansProRegular from "../assets/fonts/source-sans-pro/SourceSansPro-Regular.ttf";
import SourceSansProSemiBold from "../assets/fonts/source-sans-pro/SourceSansPro-SemiBold.ttf";

const fonts = [
  {
    "font-family": "Source Sans Pro",
    src: `url(${SourceSansProLight}) format("truetype")`,
    "font-weight": defaultLightTheme["font-weight-light"], // 300
    "font-style": "normal",
    "font-display": "swap"
  },
  {
    "font-family": "Source Sans Pro",
    src: `url(${SourceSansProRegular}) format("truetype")`,
    "font-weight": defaultLightTheme["font-weight-regular"], // 400
    "font-style": "normal",
    "font-display": "swap"
  },
  {
    "font-family": "Source Sans Pro",
    src: `url(${SourceSansProSemiBold}) format("truetype")`,
    "font-weight": defaultLightTheme["font-weight-semi-bold"], // 600
    "font-style": "normal",
    "font-display": "swap"
  }
];

const App = () => {
  return (
    <ThemeProvider
      themes={{ light: defaultLightTheme, customTheme }}
      defaultThemeName="customTheme"
      fonts={fonts}>
      {children}
    </ThemeProvider>
  );
};

render(<App />, document.getElementById("root"));
```

## Default theme

There are two default themes, `defaultLightTheme` and `defaultDarkTheme`. They are both name exported from pi-ui.

```js
import { defaultLightTheme, defaultDarkTheme } from "pi-ui";
```

## UseTheme hook in deep children

`useTheme` exposes three properties `theme`, `themeName`, `setThemeName` which can be used in deep children.

```js
import React from "react";
import { useTheme, Button, getThemeProperty } from "pi-ui";

const DeepChild = () => {
  const { themeName, setThemeName, theme, DEFAULT_LIGHT_THEME_NAME,
    DEFAULT_DARK_THEME_NAME } = useTheme();

  const colorGreen = getThemeProperty(theme, "color-green"); // get theme property 
  
  const handleToggleTheme = () => {
    if (themeName === DEFAULT_LIGHT_THEME_NAME) {
      setThemeName(DEFAULT_LIGHT_THEME_NAME);
    } else {
      setThemeName(DEFAULT_LIGHT_THEME_NAME);
    }
  }

  return (
    <Button onClick={handleToggleTheme}>Toggle theme</Button>
  )
};
```
