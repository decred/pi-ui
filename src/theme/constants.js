/*
 * The following constants are used throughout the application to determine
 * if the current active theme is default light or dark theme
 */
export const DEFAULT_LIGHT_THEME_NAME = "light";
export const DEFAULT_DARK_THEME_NAME = "dark";
