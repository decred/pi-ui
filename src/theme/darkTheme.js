import lightTheme from "./lightTheme";

const darkTheme = {
  ...lightTheme,

  "color-primary-dark": "#7DA7D9",
  "color-primary-darker": "#1F325F",
  "color-primary-darkest": "#1B2B54",
  "color-dark": "#091440",
  "color-gray-blue": "#6F98D3",

  "color-powder-blue": "#B7DEEE",
  "color-astronaut-blue": "#29417A",

  "html-bg": "var(--color-primary-darkest)",

  "text-input-color": "var(--color-gray-blue)",
  "container-color": "var(--color-primary-darkest)",
  "header-border-color": "#1d1d2f",
  "header-color": "var(--color-primary-darker)",
  "topbanner-color": "var(--card-background)",
  "main-color": "#27283d",
  "input-border-color": "var(--color-gray-blue)",
  "active-nav-border-color": "var(--color-primary-light)",
  "separator-color": "var(--color-primary-darkest)",
  "subtitle-color": "#3C62B0",
  "card-background": "#223767",
  "dimmed-card-background": "#1F2F5A",
  "success-icon-background-color": "#527CC9",
  "success-icon-checkmark-color": "var(--color-primary-darkest)",

  "btn-disabled-background-color": "var(--color-primary-darker)",
  "btn-disabled-text-color": "var(--info-message-background)",

  "comment-like-color": "var(--color-gray-blue)",
  "comment-like-color-active": "var(--color-powder-blue)",
  "comment-censored-background": "var(--dimmed-card-background)",
  "comment-new-background": "var(--color-primary-darker)",

  "text-color": "#99C1E3",
  "text-color-light": "var(--color-white)",
  "text-heading-color": "var(--color-powder-blue)",
  "text-heading2-color": "var(--text-heading-color)",
  "text-heading3-color": "#E9F8FE",
  "text-secondary-color": "var(--color-primary-dark)",

  "tab-text-color": "var(--color-gray-blue)",
  "tab-text-active-color": "#E9F8FE",
  "tab-count-background": "#2F4D8C",

  "icon-color": "#36579E",
  "icon-background-color": "var(--text-color)",
  "icon-hover-color": "var(--color-primary-dark)",

  "table-header-background": "var(--dimmed-card-background)",
  "table-cell-border-color": "var(--separator-color)",
  "table-page-arrow-background": "var(--card-background)",
  "table-page-arrow-text-color": "var(--tab-text-active-color)",
  "table-active-page-arrow-text-color": "var(--text-color)",
  "table-active-page-arrow-background": "var(--tab-count-background)",

  "md-table-cell-border-color": "var(--color-primary-dark)",
  "md-table-odd-row-background": "var(--color-primary-darker)",
  "md-table-even-row-background": "var(--color-primary-darkest)",

  "toggle-bar-color": "var(--color-primary-darkest)",

  "dropdown-item-hover-color": "var(--color-powder-blue)",
  "dropdown-arrow-color": "var(--text-color)",

  "select-anchor-color": "var(--text-color)",

  "copyable-text-background-color": "var(--color-primary-darkest)",

  "modal-close-color": "var(--text-color)",

  "tooltip-background-color": "var(--color-astronaut-blue)",

  "checkbox-stroke-color": "var(--color-gray-blue)",
  "checkbox-hovered-background-color": "var(--color-primary-darkest)",

  "info-message-background": "#152042",
  "blocked-message-background": "var(--info-message-background)",
  "error-message-background": "var(--info-message-background)",
  "error-message-color": "var(--color-astronaut-blue)",
  "blocked-message-color": "var(--error-message-color)",
  "warning-message-background": "var(--info-message-background)",
  "warning-message-color": "var(--color-astronaut-blue)",

  "datasheet-read-only-text-color": "var(--color-gray)",
  "datasheet-read-only-background": "var(--color-primary-darkest)",
  "datasheet-border-color": "var(--color-gray-blue)"
};

export default darkTheme;
