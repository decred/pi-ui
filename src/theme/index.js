export { default as defaultDarkTheme } from "./darkTheme";
export { default as defaultLightTheme } from "./lightTheme";
export { default as useFont } from "./useFont";
export * from "./ThemeProvider";
export * from "./helpers";
