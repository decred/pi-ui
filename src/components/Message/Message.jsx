import React from "react";
import PropTypes from "prop-types";
import Card from "../Card/Card.jsx";
import styles from "./styles.css";
import Icon from "../Icon/Icon.jsx";
import { classNames } from "../../utils";

const typeToIcons = {
  info: "info",
  warning: "alert",
  success: "checkmark",
  error: "alert"
};

const Message = ({ style, className, children, kind, ...props }) => {
  return (
    <Card
      className={classNames(styles[kind], className)}
      style={style}
      {...props}>
      <Icon
        type={typeToIcons[kind]}
        className={styles[`icon-${kind}`]}
        size="lg"
      />
      {children}
    </Card>
  );
};

Message.propTypes = {
  style: PropTypes.object,
  className: PropTypes.string,
  children: PropTypes.element.isRequired,
  kind: PropTypes.oneOf(["info", "warning", "error", "success"])
};
Message.defaultProps = {
  kind: "info"
};

export default Message;
