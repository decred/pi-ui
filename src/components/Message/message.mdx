---
name: Message
route: /message
---

import { Playground, PropsTable } from "docz";
import Message from "./Message.jsx";

# Message

## Properties

<PropsTable of={Message} />

## Usage

<Playground>
  <Message kind="info">Info</Message>
  <br/>
  <Message kind="warning">Warning</Message>
  <br/>
  <Message kind="error">Error</Message>
  <br/>
  <Message kind="success">Success</Message>
</Playground>
