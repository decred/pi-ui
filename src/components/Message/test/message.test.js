import React from "react";
import Message from "../Message";
import { create } from "react-test-renderer";
import { defaultLightTheme, ThemeProvider } from "../../../theme";

describe("Message component", () => {
  test("Matches the snapshot", () => {
    const message = create(
      <ThemeProvider
        themes={{ light: defaultLightTheme }}
        defaultThemeName="light">
        <Message>test</Message>
      </ThemeProvider>
    );
    expect(message.toJSON()).toMatchSnapshot();
  });
});
