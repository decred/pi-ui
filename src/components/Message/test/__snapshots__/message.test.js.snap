// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message component Matches the snapshot 1`] = `
<div
  className="card padding-undefined info"
>
  <div
    className="iconContainer"
  >
    <svg
      className="icon-info"
      height="20"
      viewBox="0 0 450 450"
      width="20"
    >
      <path
        d="M930.782,3.4C808.266,3.4,705.8,105.868,705.8,228.383S808.266,453.367,930.782,453.367,1155.766,350.9,1155.766,228.383C1157.991,108.1,1053.3,3.4,930.782,3.4Zm24.5,372H908.5V174.922h46.778Zm-24.5-236.121c-17.819,0-28.959-11.138-28.959-26.731S915.188,85.82,930.782,85.82c17.821,0,28.959,11.138,28.959,26.731C961.967,128.143,948.6,139.281,930.782,139.281Z"
        fill="#c4cbd2"
        transform="translate(-705.799 -3.4)"
      />
    </svg>
  </div>
  <div
    className="content"
  >
    test
  </div>
</div>
`;
